### ISSUE TITLE
Mixin issues

### ISSUE THREAD

**Issue by raulsebastianmihaila** [michaelficarra/proposal-first-class-protocols#29]
I think that the fact that protocols are also a kind of mixins is problematic. First, because they are a naive kind of mixins and, secondly, because I think it would be better to separate the two mechanisms. These are the issues with this kind of mixins:

- the gorilla-banana problem: you get everything the protocol contains even though you might not need everything
- property collisions: the fact that the first mixin always wins (in case the property is not overwritten by the class) isn't really a solution
- when looking at a class definition you can't tell what methods you have/inherit.
- when you see an inherited method in one of your own methods, you can't tell where it's coming from (you have to look at all the protocols you're implementing potentially)
- you can not share private state. sharing private state is useful when you're reusing code to define a concept that has private state

I have a [proposal](https://github.com/raulsebastianmihaila/ecma262-definition-mixins-proposal) for a different kind of mixins that don't have all these issues. They also work with both function syntax and class syntax and I believe they can be made to work with protocols. This way the protocol and mixin mechanisms can stay separate.

Please see [here](https://esdiscuss.org/topic/definition-mixins) a more detailed ongoing discussion.

**Comment by Jamesernator** [comment]
Despite appearances the protocols proposed aren't using strings by default for example this:

```js
protocol Fizz {
    buzz;
}

class Foo extends Fizz {
    buzz() {

    }
}
```

Is not a valid implementation of the protocol `Fizz` (in fact it throws an error for being not implemented), `Fizz.buzz` is actually a `Symbol` not a string so in order to implement it you need to use:

```js
class Foo extends Fizz {
    [Fizz.buzz]() {

    }
}

new Foo().buzz // undefined
new Foo()[Fizz.buzz] // [Function: Fizz.buzz]
Fizz.buzz // Symbol("Fizz.buzz")
```

With this most of your points don't apply (unless using the string based version which is there for backwards compatibility with existing protocols not necessarily for general use):

- Property collisions can't happen because the protocol symbols are unique
- Similarly looking at a class definition you definitely know what methods are inherited from the mixi
- And as consequence you always know what you inherit as methods must be accessed via the Symbol keys

Now I don't really have suggestions for the gorilla-banana problem other than to just not worry about it, if your protocols have lots of fields then this might be a design flaw with the protocol but if it isn't then there's surely a reason the protocol requires those fields? Something that is probably needed is optional fields when not everything is required to be implemented e.g. maybe something like (not sure if this is what you were getting at):

```js
// Supposing Iterator was implemented using the proposed protocol
protocol Iterator {
    next;
    return?; // Still creates Iterator.return 
             // but doesn't require implementation
    throws?; // Ditto
}

class ArrayIterator extends Iterator {
    [Iterator.next]() { /* Not optional */ }
    [Iterator.return]() { /* Optional, could be emitted */ }
}
```

And regarding the private state thing, I raised [a point](https://github.com/michaelficarra/proposal-first-class-protocols/issues/15#issuecomment-329325429) about this a while ago, however I don't think it's super critical at least in the initial designs of this proposal given that it's generally bad practice to poke and mess with arbitrary symbol properties on objects.

**Comment by raulsebastianmihaila** [comment]
> Property collisions can't happen because the protocol symbols are unique
In fact, the same symbol can be used in different protocols. See this [issue](https://github.com/michaelficarra/proposal-first-class-protocols/issues/20)

> Similarly looking at a class definition you definitely know what methods are inherited from the mixin
What do you mean? You can not do that if you inherit the implementations. You have to look in the protocols themselves.

> And as consequence you always know what you inherit as methods must be accessed via the Symbol keys
That is only if you implement the actual method yourself.

All these, together with sharing private state, are probably not important issues for protocols since probably protocols will mostly not be used as mixins, although they could easily be abused like that. My point was that since they can work as mixins, it would be better to have a more capable and better designed mixin mechanism and use that for protocols as well.

**Comment by Jamesernator** [comment]
> In fact, the same symbol can be used in different protocols. See this issue

Yes but this is ultimately a design decision, reusing a Symbol (inside a protocol or not) risks collision. I'd have to see a concrete example of the problem in #20 to see that you'd ever want to do this in practice in a way that would cause collisions.

> What do you mean? You can not do that if you inherit the implementations. You have to look in the protocols themselves.

Well I mean you know that it inherits everything that the protocol defines, by definition you already have access to the protocol to implement it so given that fact you can simply use whatever is available. There's nothing different from subclassing here, you might need to inspect the chain upwards to discover all things a class has I don't see anyway around this e.g.:

```js
const bar = Symbol('bar')

class A {
   foo() {

   }

   [bar]() {}
}

protocol B {
   fizz() {
       return 12
   }
}

// For both A and B you simply have to inspect their definition to know
// what you're getting, there's nothing special about protocol in this
// regard, you simply have to inspect the rest of the chain
class C extends A implements B {}
```

> That is only if you implement the actual method yourself.

I don't understand what you mean, pre-defined ones will simply exist on the thing that implements the protocol:

```js
protocol Mixin {
    mixinMethod() {
        return 10;
    }
}

class Foo implements Mixin {}

new Foo()[Mixin.mixinMethod]() // 10

if (someArbitraryValue implements Mixin) {
    // If something implements a protocol it gets
    // pre-defined methods for free
    someArbitraryValue[Mixin.mixinMethod]() // 10
}
```

**Comment by raulsebastianmihaila** [comment]
> Yes but this is ultimately a design decision, reusing a Symbol (inside a protocol or not) risks collision. I'd have to see a concrete example of the problem in #20 to see that you'd ever want to do this in practice in a way that would cause collisions.

I was just saying you can get collisions, I'm saying it's ok to share the symbols.

>Well I mean you know that it inherits everything that the protocol defines, by definition you already have access to the protocol to implement it so given that fact you can simply use whatever is available. There's nothing different from subclassing here, you might need to inspect the chain upwards to discover all things a class has I don't see anyway around

Sure, inheritance has this issue as well. In fact, this kind of mixins are a sort of multiple inheritance.

Again, my point is that if later a generic well designed mixin mechanism is added later (and I wish it was sooner), it would be better for the protocols to use it instead of having yet another kind of inheritance.

**Comment by michaelficarra** [comment]
@raulsebastianmihaila What is it that you would like out of a "mixin" feature that you don't feel protocols provide? I would like for this proposal to satisfy the community's desires for mixins. The way I understand it, mixins are simply protocols with no required symbols. All of the points in your original comment have been addressed, and the "gorilla-banana problem" you're referring to doesn't seem to cause an actual issue: there's no downside to having additional symbol-named properties.

**Comment by raulsebastianmihaila** [comment]
I missed the fact that protocols can not provide default implementations for properties with string keys. Sorry for that. That means that property collisions are indeed very likely to happen with protocols. However this also makes them less feasible as a general mixin mechanism, as symbols are more difficult to work with than strings.

**Comment by aluanhaddad** [comment]
Without the ability to implement string keys, protocols cannot serve the purpose of general mixins.

The detractions to ergonomics, discoverability, and transparency implied by Symbol keys will make using protocols as mixins DOA.

However, from reading the recent meeting notes it seems that protocols are intended to subsume and mixins.

**Comment by michaelficarra** [comment]
@aluanhaddad How would you expect a mixin system based on string-named properties to handle collisions?
